musmart/melody/contour/step_contour.py

```diff
@@ -0,0 +1,217 @@
+        each pitch value proportionally to its normalized duration.
+
+        Parameters
+        ----------
+        pitches : list[int]
+            List of pitch values
+        normalized_durations : list[float]
+            List of normalized duration values (should sum to step_contour_length)
+
+        Returns
+        -------
+        list[int]
+            List of length step_contour_length containing repeated pitch values
+
+        Examples
+        --------
+        >>> sc = StepContour([60], [1.0], step_contour_length=4)
+        >>> sc._expand_to_vector([60, 62], [2.0, 2.0])
+        [60, 60, 62, 62]
+        """
+        result = []
+        for pitch, duration in zip(pitches, normalized_durations):
+            repetitions = round(duration)
+            result.extend([pitch] * repetitions)
+
+        if len(result) > self._step_contour_length:
+            result = result[:self._step_contour_length]
+        elif len(result) < self._step_contour_length:
+            result.extend([result[-1]] * (self._step_contour_length - len(result)))
+
+        return result
+
+    def _calculate_contour(
+        self,
+        pitches: list[int],
+        durations: list[float]
+    ) -> list[int]:
+        """Calculate the step contour from input pitches and durations.
+
+        Examples
+        --------
+        >>> sc = StepContour([60, 62], [1.0, 1.0], step_contour_length=4)
+        >>> sc._calculate_contour([60, 62], [1.0, 1.0])
+        [60, 60, 62, 62]
+        """
+        normalized_durations = self._normalize_durations(durations)
+        return self._expand_to_vector(pitches, normalized_durations)
+
+    @property
+    def global_variation(self) -> float:
+        """Calculate the global variation of the step contour.
+
+        Returns
+        -------
+        float
+            Float value representing the global variation of the step contour
+
+        Examples
+        --------
+        >>> sc = StepContour([60, 62, 64], [1.0, 1.0, 1.0])
+        >>> round(sc.global_variation, 2)
+        1.64
+        """
+        return float(np.nanstd(self.contour))
+
+    @property
+    def global_direction(self) -> float:
+        """Calculate the global direction of the step contour.
+
+        Returns
+        -------
+        float
+            Float value representing the global direction of the step contour
+            Returns 0.0 if the contour is flat
+
+        Examples
+        --------
+        >>> sc = StepContour([60, 62, 64], [1.0, 1.0, 1.0])
+        >>> round(sc.global_direction, 3)
+        0.943
+        """
+        if len(set(self.contour)) == 1:
+            return 0.0
+        corr = np.corrcoef(
+            self.contour,
+            np.arange(self._step_contour_length)
+        )[0, 1]
+        return float(corr)
+
+    @property
+    def local_variation(self) -> float:
+        """Calculate the local variation of the step contour.
```

Comment:
Need to add mathematical definition of this

Reply:
done